### ISSUE TITLE
Unable to search for Chinese Keywords

### ISSUE THREAD

**Issue by taills** [rust-lang-nursery/mdBook#1081]
Unable to search for Chinese Keywords
![image](https://user-images.githubusercontent.com/47925925/67736356-4857a600-fa42-11e9-9cbb-f86b98555f45.png)

**Comment by weihanglo** [comment]
mdBook uses [elasticlunr.js](https://github.com/weixsong/elasticlunr.js) for offline searching under the hood. And according to this issue https://github.com/weixsong/elasticlunr.js/issues/53, it seems that there is no plan to support searching in other languages.

**Comment by hysencn** [comment]
it is a good tools. I love it.

We, more than 1 million peoples, have the same issue. could you help on it?
need chinese search support.

or, could we search chinese with google? 
How to do it?

**Comment by futurist** [comment]
For Chinese like languages, maybe it's not suitable for local search, instead could using some service like [algolia](https://github.com/algolia/docsearch), which also being used in [vuepress](https://github.com/vuejs/vuepress/blob/master/packages/%40vuepress/theme-default/components/AlgoliaSearchBox.vue)

**Comment by cxumol** [comment]
Yes, it is highly possible to add searching in Chinese characters.   

1. **elasticlunr**  is the search engine used by mdBook
2. Go to [elasticlunr official documentation](http://elasticlunr.com/docs/index.html), read the section _Other Languages_. With **just 3 more lines of code**, elasticlunr can be used with other languages.
3. The Chinese language support of lunr-languages is [PRed but not yet merged](https://github.com/MihaiValentin/lunr-languages/pull/53). 
4. Alternatively, suggested by comments on https://github.com/MihaiValentin/lunr-languages/issues/32, Japanese language can be used as a **workaround**. That's because [this line](https://github.com/MihaiValentin/lunr-languages/blob/a62fec97fb1a62bb4581c9b69a5ddedf62f8f62f/lunr.ja.js#L169) has covered `一-龠`, which is a usual range including most Chinese characters on the Unicode table.

**Comment by bigdogs** [comment]
Any new progress for this issue?

**Comment by Sciroccogti** [comment]
> Any new progress for this issue?

There is a PR #1496 working on it but needs help.

**Comment by wc7086** [comment]
Replacing `elasticlunr.js` with https://github.com/ajitid/fzf-for-js may allow this issue to be resolved.

**Comment by silence-coding** [comment]
Looking forward to new progress on this MR

**Comment by switefaster** [comment]
Importing `lunr.zh.min.js`(with slight modification to make it compatible with `elasticlunr`) and relevant extensions never works as mdBook is using a pre-generated search index from [elasticlunr-rs](https://github.com/mattico/elasticlunr-rs) when building the book. The correct way to solve this is PR #1496. However, the original PR is not using a preferable solution besides some flaws pointed out by the core maintainers, for example, it is not using an appropriate segmenter. And also, the PR is seemingly not updating anymore. I've figured out a possibly more elegant solution, namely to use either [Intl.Segmenter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter) or [jieba-wasm](https://www.npmjs.com/package/jieba-wasm) as a Chinese segmenter. I'd love to work on this issue, but as per the [Contribution Guide](https://github.com/rust-lang/mdBook/blob/master/CONTRIBUTING.md), I'm not sure if this issue is grabbing any attention from the maintainers, and I won't bother making a pull request if they do not.

Apart from all these, there are still some details we need to discuss:
- Whether to use `Intl.Segmenter` or `jieba-wasm`. I prefer the latter since `Intl.Segmenter` is not supported by FireFox while WebAssembly is fully supported by almost all browsers, and using `jieba` can assure the consistency between the generated index and segment generated in the browser. However, using `jieba-wasm` requires extra file dependencies to at least two files `jieba_rs_wasm.js` and `jieba_rs_wasm_bg.wasm`, I'm not sure if the maintainers will be happy with that even if we can control them with feature flags.
- `elasticlunr-rs`'s [Chinese support](https://github.com/mattico/elasticlunr-rs/blob/29d97e4c8e91bb0d1813716fb2d1575066344d76/src/lang/zh.rs#L44) is incomplete, in the way that the stop word filter is inconsistent with the one from [lunr-languages](https://github.com/MihaiValentin/lunr-languages/blob/f313734d145048be2f3681b756f9bf925aa299a1/lunr.zh.js#L142), and there is no sign of it being solved. We can implement [Language trait](https://docs.rs/elasticlunr-rs/latest/elasticlunr/lang/trait.Language.html) ourselves in mdBook, but again I'm not sure if this is appropriate as it is kind of irrelevant to mdBook itself. If the maintainers prefer not, we then have to make a PR to `elasticlunr-rs` first and then wait for it to be merged.
- ...and more if any

I'll be keeping an eye on this comment to see if anyone is interested.

**Comment by chens08** [comment]
How to use it，can you give a eaxmple about the book.toml?

**Comment by miaomiao1992** [comment]
Any new progress on this issue?

**Comment by aneteanetes** [comment]
> **just 3 more lines of code**

It was a lie. 

Аnd kind of long journey...

But anyway, here am i, and my instructions for adding non-english search by little blood:
1. First of all, you need add [lunr.stemmer.support.js](https://github.com/weixsong/lunr-languages/blob/master/lunr.stemmer.support.js) and [lunr.YOURLANG.js](https://github.com/weixsong/lunr-languages/blob/master/lunr.ru.js). You can do this by multiple ways:
 1.1 Create `head.hbs` in `theme` folder and add html-tag
 1.2 Add scripts by `additional-js` key in `book.toml`
 1.3 Or just append this script's to overrided file in step 4
3. Additionally, i am advice add [lunr.multi.js](https://github.com/weixsong/lunr-languages/blob/master/lunr.multi.js) too.
4. Then, you need `override searcher.js` by putting copy of original file into `src` folder.
5. The best part. Find `searchindex = elasticlunr.Index.load(config.index);` line, and **replace** with:
```javascript
searchindex = elasticlunr(function() {
            // adding (multi)language
            this.use(elasticlunr.multiLanguage('en', 'ru')); 
            
            // fields to index.
            this.addField('title');
            this.addField('body');
            this.addField('breadcrumbs');

            // Identify documents field
            this.setRef('id');
            
            // Get all documents stored in prebuilded index
            for (let key in config.index.documentStore.docs) {
              this.addDoc(config.index.documentStore.docs[key]);
            }
        });
```

And search will be fine.

But one more word: when mdbook build `searchindex`, it's completely ignore language settings in `book.toml`. It's uses only english chars for creating index, while [elasticlunr-rs](https://docs.rs/elasticlunr-rs/latest/elasticlunr/#elasticlunr-rs) support other languages. By this behaviour, all attempts for adding additinal language will fail. I do not write Rust code, and can't create PR, but i hope this information will help someone.

**Comment by luizbgomide** [comment]
This issue is 5 years old and no actuall progress in the code base yet? It fails even with diacrits like áàã, etc witch in theory could be very easy to handle with normalization and canonical decomposition...
